File: src/main/kotlin/ArzFile.cs
// Decompiled with JetBrains decompiler
// Type: TQ_weaponsmith.ArzFile
// Assembly: TQ_Blacksmith, Version=0.1.4.1, Culture=neutral, PublicKeyToken=null
// MVID: 1CC9040D-95C6-4C54-B248-9D80684B2CA8
// Assembly location: D:\Games\GOG Galaxy\Games\Titan Quest - Anniversary Edition\Downloads\Blacksmith\TQ_Blacksmith.exe

using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.IO.Compression;

namespace TQ_weaponsmith
{
  public class ArzFile
  {
    private byte[] rawData;
    private string fileName;
    private string[] strings;
    private Dictionary<string, ArzFile.RecordInfo> recordInfo;
    private Dictionary<string, DBRecord> cache;
    private string[] keys;
    private string[] requiredKeys;
    public int requiredKeyIndex;

    public ArzFile(string fileName)
    {
      this.fileName = fileName;
      this.cache = new Dictionary<string, DBRecord>();
    }

    private void BuildKeyTable()
    {
      if (this.recordInfo == null || this.recordInfo.Count == 0)
        return;
      int index = 0;
      DBClasses.getClasses();
      this.keys = new string[this.recordInfo.Count];
      this.requiredKeys = new string[this.recordInfo.Count];
      this.requiredKeyIndex = 0;
      foreach (string key in this.recordInfo.Keys)
      {
        this.keys[index] = key;
        string recordType = this.recordInfo[key].RecordType;
        if (recordType != null && !string.IsNullOrEmpty(recordType))
          this.requiredKeys[this.requiredKeyIndex++] = key;
        ++index;
      }
      Array.Sort<string>(this.keys);
    }

    public DBRecord GetItem(string recordId)
    {
      DBRecord dbRecord = (DBRecord) null;
      if (string.IsNullOrEmpty(recordId))
        return dbRecord;
      recordId = TQData.NormalizeRecordPath(recordId);
      try
      {
        return this.cache[recordId];
      }
      catch (KeyNotFoundException ex1)
      {
        ArzFile.RecordInfo recordInfo;
        try
        {
          recordInfo = this.recordInfo[recordId];
        }
        catch (KeyNotFoundException ex2)
        {
          return (DBRecord) null;
        }
        return recordInfo.Decompress(this);
      }
    }

    public string[] GetKeyTable()
    {
      if (this.keys == null || this.keys.Length == 0)
        this.BuildKeyTable();
      return (string[]) this.keys.Clone();
    }

    public string[] GetRequiredKeyTable()
    {
      if (this.requiredKeys == null || this.requiredKeys.Length == 0)
        this.BuildKeyTable();
      return (string[]) this.requiredKeys.Clone();
    }

    public DBRecord GetRecordNotCached(string recordId)
    {
      recordId = TQData.NormalizeRecordPath(recordId);
      try
      {
        return this.cache[recordId];
      }
      catch (KeyNotFoundException ex1)
      {
        try
        {
          return this.recordInfo[recordId].Decompress(this);
        }
        catch (KeyNotFoundException ex2)
        {
          return (DBRecord) null;
        }
      }
    }

    private string Getstring(int index) => this.strings[index];

    public bool Read()
    {
      StreamWriter outStream = (StreamWriter) null;
      if (TQDebug.DatabaseDebugLevel > 2)
        outStream = new StreamWriter("arzOut.txt", false);
      try
      {
        FileStream input = new FileStream(this.fileName, FileMode.Open, FileAccess.Read, FileShare.Read);
        this.rawData = new BinaryReader((Stream) input).ReadBytes((int) input.Length);
        BinaryReader reader = new BinaryReader((Stream) new MemoryStream(this.rawData, false));
        try
        {
          int[] numArray = new int[6];
          for (int index = 0; index < 6; ++index)
          {
            numArray[index] = reader.ReadInt32();
            outStream?.WriteLine("Header[{0}] = {1:n0} (0x{1:X})", (object) index, (object) numArray[index]);
          }
          int pos = numArray[1];
          int numEntries = numArray[3];
          this.ReadStringTable(numArray[4], reader, outStream);
          this.ReadRecordTable(pos, numEntries, reader, outStream);
          for (int index = 0; index < 4; ++index)
          {
            int num = reader.ReadInt32();
            outStream?.WriteLine("{0:n0} 0x{0:X}", (object) num);
          }
        }
        catch (IOException ex)
        {
          throw;
        }
        finally
        {
          reader.Close();
        }
      }
      catch (IOException ex)
      {
        if (!TQDebug.DebugEnabled)
          TQDebug.DebugEnabled = true;
        TQDebug.DebugWriteLine(ex.ToString());
        return false;
      }
      finally
      {
        outStream?.Close();
      }
      return true;
    }

    private void ReadRecordTable(
      int pos,
      int numEntries,
      BinaryReader reader,
      StreamWriter outStream)
    {
      this.recordInfo = new Dictionary<string, ArzFile.RecordInfo>((int) Math.Round((double) numEntries * 1.2));
      reader.BaseStream.Seek((long) pos, SeekOrigin.Begin);
      outStream?.WriteLine("RecordTable located at 0x{0:X}", (object) pos);
      for (int index = 0; index < numEntries; ++index)
      {
        ArzFile.RecordInfo recordInfo = new ArzFile.RecordInfo();
        recordInfo.Decode(reader, 24, this);
        this.recordInfo.Add(TQData.NormalizeRecordPath(recordInfo.ID), recordInfo);
        outStream?.WriteLine("{0},{1},{2}", (object) index, (object) recordInfo.ID, (object) recordInfo.RecordType);
      }
    }

    private void ReadStringTable(int pos, BinaryReader reader, StreamWriter outStream)
    {
      reader.BaseStream.Seek((long) pos, SeekOrigin.Begin);
      int length = reader.ReadInt32();
      this.strings = new string[length];
      outStream?.WriteLine("stringTable located at 0x{1:X} numstrings= {0:n0}", (object) length, (object) pos);
      for (int index = 0; index < length; ++index)
      {
        this.strings[index] = TQData.ReadCString(reader);
        outStream?.WriteLine("{0},{1}", (object) index, (object) this.strings[index]);
      }
    }

    public int Count => this.recordInfo.Count;

    public class RecordInfo
    {
      private int offset;
      private int idStringIndex = -1;
      public string ID;
      public string RecordType;

      public RecordInfo() => this.RecordType = string.Empty;

      public void Decode(BinaryReader inReader, int baseOffset, ArzFile arzFile)
      {
        this.idStringIndex = inReader.ReadInt32();
        this.RecordType = TQData.ReadCString(inReader);
        this.offset = inReader.ReadInt32() + baseOffset;
        inReader.ReadInt32();
        inReader.ReadInt32();
        inReader.ReadInt32();
        this.ID = arzFile.Getstring(this.idStringIndex);
      }

      public DBRecord Decompress(ArzFile arzFile)
      {
        byte[] buffer = this.DecompressBytes(arzFile);
        int num1 = buffer.Length / 4;
        if (buffer.Length % 4 > 0)
        {
          if (!TQDebug.DebugEnabled)
            TQDebug.DebugEnabled = true;
          TQDebug.DebugWriteLine(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error in ARZFile - {0}", new object[1]
          {
            (object) arzFile.fileName
          }));
          TQDebug.DebugWriteLine(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error while parsing arz record {0}, data Length = {1} which is not a multiple of 4", new object[2]
          {
            (object) this.ID,
            (object) buffer.Length
          }));
          throw new ArgumentException(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error while parsing arz record {0}, data Length = {1} which is not a multiple of 4", new object[2]
          {
            (object) this.ID,
            (object) buffer.Length
          }));
        }
        DBRecord dbRecord = new DBRecord(this.ID, this.RecordType);
        using (BinaryReader binaryReader = new BinaryReader((Stream) new MemoryStream(buffer, false)))
        {
          int num2 = 0;
          while (num2 < num1)
          {
            short dataType = binaryReader.ReadInt16();
            short numberOfValues = binaryReader.ReadInt16();
            int index1 = binaryReader.ReadInt32();
            string variableName = arzFile.Getstring(index1);
            if (variableName == null)
            {
              if (!TQDebug.DebugEnabled)
                TQDebug.DebugEnabled = true;
              TQDebug.DebugWriteLine(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error in ARZFile - {0}", new object[1]
              {
                (object) arzFile.fileName
              }));
              TQDebug.DebugWriteLine(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error while parsing arz record {0}, variable is NULL", new object[1]
              {
                (object) this.ID
              }));
              throw new ArgumentNullException(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error while parsing arz record {0}, variable is NULL", new object[1]
              {
                (object) this.ID
              }));
            }
            if (dataType < (short) 0 || dataType > (short) 3)
            {
              if (!TQDebug.DebugEnabled)
                TQDebug.DebugEnabled = true;
              TQDebug.DebugWriteLine(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error in ARZFile - {0}", new object[1]
              {
                (object) arzFile.fileName
              }));
              TQDebug.DebugWriteLine(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error while parsing arz record {0}, variable {2}, bad dataType {3}", new object[3]
              {
                (object) this.ID,
                (object) variableName,
                (object) dataType
              }));
              throw new ArgumentOutOfRangeException(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error while parsing arz record {0}, variable {2}, bad dataType {3}", new object[3]
              {
                (object) this.ID,
                (object) variableName,
                (object) dataType
              }));
            }
            Variable variable = new Variable(variableName, (VariableDataType) dataType, (int) numberOfValues);
            if (numberOfValues < (short) 1)
            {
              if (!TQDebug.DebugEnabled)
                TQDebug.DebugEnabled = true;
              TQDebug.DebugWriteLine(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error in ARZFile - {0}", new object[1]
              {
                (object) arzFile.fileName
              }));
              TQDebug.DebugWriteLine(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error while parsing arz record {0}, variable {1}, bad valCount {2}", new object[3]
              {
                (object) this.ID,
                (object) variableName,
                (object) numberOfValues
              }));
              throw new ArgumentException(string.Format((IFormatProvider) CultureInfo.InvariantCulture, "Error while parsing arz record {0}, variable {1}, bad valCount {2}", new object[3]
              {
                (object) this.ID,
                (object) variableName,
                (object) numberOfValues
              }));
            }
            num2 += 2 + (int) numberOfValues;
            for (int index2 = 0; index2 < (int) numberOfValues; ++index2)
            {
              switch (variable.DataType)
              {
                case VariableDataType.Integer:
                case VariableDataType.Boolean:
                  int num3 = binaryReader.ReadInt32();
                  variable[index2] = (object) num3;
                  break;
                case VariableDataType.Float:
                  float num4 = binaryReader.ReadSingle();
                  variable[index2] = (object) num4;
                  break;
                case VariableDataType.StringVar:
                  int index3 = binaryReader.ReadInt32();
                  string str1 = arzFile.Getstring(index3);
                  string str2 = str1 != null ? str1.Trim() : string.Empty;
                  variable[index2] = (object) str2;
                  break;
                default:
                  int num5 = binaryReader.ReadInt32();
                  variable[index2] = (object) num5;
                  break;
              }
            }
            dbRecord.Set(variable);
          }
        }
        return dbRecord;
      }

      private byte[] DecompressBytes(ArzFile arzFile)
      {
        if (arzFile == null)
          throw new ArgumentNullException(nameof (arzFile), "arzFile is null.");
        using (MemoryStream memoryStream1 = new MemoryStream(arzFile.rawData, false))
        {
          memoryStream1.Seek((long) this.offset, SeekOrigin.Begin);
          memoryStream1.ReadByte();
          memoryStream1.ReadByte();
          using (DeflateStream deflateStream = new DeflateStream((Stream) memoryStream1, CompressionMode.Decompress))
          {
            using (MemoryStream memoryStream2 = new MemoryStream())
            {
              byte[] buffer = new byte[1024];
              int count;
              while ((count = deflateStream.Read(buffer, 0, 1024)) > 0)
                memoryStream2.Write(buffer, 0, count);
              return memoryStream2.ToArray();
            }
          }
        }
      }
    }
  }
}

File: src/main/kotlin/ArzFile.kt
import utils.LittleEndianDataInputStream
import java.io.BufferedInputStream
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.DataInputStream
import java.io.InputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.FileChannel
import java.nio.file.Files
import java.util.zip.Inflater
import kotlin.math.roundToInt


private val logger = mu.KotlinLogging.logger {}
public class  ArzFile constructor(filename:String) {
  private val fileName: String

  //    private var rawData = ByteBuffer.allocate(0)
  private val strings = mutableListOf<String>()
  private var recordInfo = mutableMapOf<String, ArzFile.RecordInfo>()
  private val cache = mutableMapOf<String, DBRecord>()
  private val keys = mutableListOf<String>()
  private val requiredKeys = mutableListOf<String>()
  public var requiredKeyIndex = 0

  init {
    this.fileName = filename
  }

  private fun BuildKeyTable(): Unit {
    if (this.recordInfo == null || this.recordInfo.size == 0) {
      return;
    }
    var index = 0;
//      DBClasses.getClasses();
    this.requiredKeyIndex = 0;
    recordInfo.keys.sorted().forEach { key ->

      this.keys[index] = key;
      val recordType = this.recordInfo[key]?.RecordType;
      if (recordType != null && !recordType.isNullOrEmpty())
        this.requiredKeys[this.requiredKeyIndex++] = key
      index++
    }
  }

  public fun GetItem(recordIdOrg: String): DBRecord? {
    if (recordIdOrg.isNullOrEmpty()) {
      return null
    }
    val recordId = TQData.NormalizeRecordPath(recordIdOrg);
    return if (cache.containsKey(recordId)) {
      this.cache[recordId];
    } else {
      if (recordInfo.containsKey(recordId)) {
        val recordInfo = this.recordInfo[recordId]
        recordInfo?.Decompress(this)
      } else {
        null
      }

    }
  }

  public fun GetKeyTable(): List<String> {
    if (this.keys == null || this.keys.size == 0)
      this.BuildKeyTable();
    return this.keys.toList()
  }

  public fun GetRequiredKeyTable(): List<String> {
    if (this.requiredKeys == null || this.requiredKeys.size == 0)
      this.BuildKeyTable();
    return this.requiredKeys.toList()
  }

  public fun GetRecordNotCached(recordIdOrg: String): DBRecord? {
    val recordId = TQData.NormalizeRecordPath(recordIdOrg);
    return if (cache.containsKey(recordId)) {
      this.cache[recordId]
    } else {
      if (recordInfo.containsKey(recordId)) {
        this.recordInfo[recordId]?.Decompress(this)
      } else {
        null
      }
    }
  }

  private fun Getstring(index: Int) = this.strings[index].orEmpty()

  public fun Read(): Boolean {


    val file = File(fileName)

    FileChannel.open(file.toPath()).use {
      val reader = it.map(FileChannel.MapMode.READ_ONLY, 0, file.length())
      reader.order(ByteOrder.LITTLE_ENDIAN)
//        rawData = reader

//        reader.get(rawData)
//      val reader = LittleEndianDataInputStream(BufferedInputStream(file.inputStream()))
//      rawData = Files.readAllBytes(file.toPath())
      reader.position(0)

      val numArray = IntArray(6)
      for (index in numArray.indices) {
        val header = reader.getInt()
        numArray[index] = header
        val hex = "(0x%01X)".format(header)
        logger.info { "Header[${index}] = ${header} $hex" }
      }
      val pos = numArray[1]
      val numEntries = numArray[3];
      this.ReadStringTable(numArray[4], reader);
      this.ReadRecordTable(pos, numEntries, reader);
      for (index in 0 until 4) {
        val num = reader.getInt()
        val hex = "$num 0x%01X".format(num)
        logger.debug { "Footer $hex" }
      }
    }
//
//      catch ( ex: IOException)
//      {
//        if (!TQDebug.DebugEnabled)
//          TQDebug.DebugEnabled = true;
//        TQDebug.DebugWriteLine(ex.ToString());
//        return false;
//      }
//      finally
//      {
//        outStream?.Close();
//      }
    return true;
  }

  private fun ReadRecordTable(
    pos: Int,
    numEntries: Int,
    reader: ByteBuffer
  ) {
    this.recordInfo = LinkedHashMap((numEntries * 1.2).roundToInt());
    reader.position(pos)
    logger.info { "RecordTable located at $pos" }
    for (index in 0 until numEntries) {
      val recordInfo = ArzFile.RecordInfo();
      recordInfo.Decode(reader, 24, this);
      this.recordInfo.put(TQData.NormalizeRecordPath(recordInfo.ID), recordInfo);
      logger.debug { "recordtable $index ${recordInfo.ID} ${recordInfo.RecordType}" }
    }
  }

  private fun ReadStringTable(pos: Int, reader: ByteBuffer) {
    reader.position(pos)
    val length = reader.getInt()
    logger.debug { "stringTable located at $pos numstrings= $length" }
    for (index in 0 until length) {
      this.strings.add(index, TQData.readCString(reader))
      logger.debug { "stringtable entry $index ${strings[index]}" }
    }
  }

  public fun Count() = this.recordInfo.size

  public class RecordInfo {
    private var offset = 0
    private var idStringIndex = -1
    public var ID = ""
    public var RecordType = ""


    public fun Decode(inReader: ByteBuffer, baseOffset: Int, arzFile: ArzFile) {
      this.idStringIndex = inReader.getInt()
      this.RecordType = TQData.readCString(inReader);
      this.offset = inReader.getInt() + baseOffset;
      inReader.getInt()
      inReader.getInt()
      inReader.getInt()
      this.ID = arzFile.Getstring(this.idStringIndex);
    }

    public fun Decompress(arzFile: ArzFile): DBRecord? {
      val buffer = this.DecompressBytes(arzFile)
      val num1 = buffer.size / 4;
      if (buffer.size % 4 > 0) {

        logger.error { "Error in ARZFile - ${arzFile.fileName} \n Error while parsing arz record ${this.ID}, data Length = ${buffer.size} which is not a multiple of 4" }
        return null
      }
      val dbRecord = DBRecord(this.ID, this.RecordType)
      LittleEndianDataInputStream(ByteArrayInputStream(buffer)).use {
        var num2 = 0
        while (num2 < num1) {
          val dataType = it.readShort().toInt()
          val numberOfValues = it.readShort().toInt()
          val index1 = it.readInt()
          val variableName = arzFile.Getstring(index1);
          if (variableName.isEmpty()) {
            logger.error { "Error in ARZFile - ${arzFile.fileName} \nError while parsing arz record ${ID}, variable is NULL" }
            return null
          }
          if (dataType < 0 || dataType > 3) {
            logger.error { "Error in ARZFile - ${arzFile.fileName}\nError while parsing arz record $ID, variable $variableName, bad dataType $dataType" }
            return null
          }
          val variable = Variable(variableName, VariableDataType.values()[dataType], numberOfValues);
          if (numberOfValues < 1) {
            logger.error { "Error in ARZFile - ${arzFile.fileName}\nError while parsing arz record $ID, variable $variableName, bad valCount $numberOfValues" }
            return null
          }
          num2 += 2 + numberOfValues
          for (index2 in 0 until numberOfValues) {
            when (variable.DataType) {
              VariableDataType.Integer,
              VariableDataType.Boolean -> {
                val num3 = it.readInt()
                variable.setString(index2, num3.toString())
              }

              VariableDataType.Float -> {
                val num4 = it.readFloat()
                variable.setString(index2, num4.toString())
              }

              VariableDataType.StringVar -> {
                val index3 = it.readInt()
                val str1 = arzFile.Getstring(index3);
                variable.setString(index2, str1.trim())
              }

              else -> {
                val num5 = it.readInt()
                variable.setString(index2, num5.toString())
              }
            }
            dbRecord.Set(variable);
          }
        }
      }
      return dbRecord;
    }

    private fun DecompressBytes(arzFile: ArzFile): ByteArray {
      val inflater = Inflater()
      val file = File(arzFile.fileName)

      FileChannel.open(file.toPath()).use {
        val map = it.map(FileChannel.MapMode.READ_ONLY, offset.toLong(), file.length() - offset)
        map.order(ByteOrder.LITTLE_ENDIAN)
        inflater.setInput(map)
//          memoryStream1.read()
//          memoryStream1.read()
        val memoryStream2 = ByteArrayOutputStream()
        memoryStream2.use {
          val buffer = ByteArray(1024)
          var count = 0
          while (!inflater.finished()) {
            count = inflater.inflate(buffer)
            memoryStream2.write(buffer, 0, count)
          }
        }
        return memoryStream2.toByteArray()
      }
    }
  }
}




